### PR TITLE
refactor: separate processing of user-defined steps and teardown steps

### PR DESCRIPTION
**Pull request by FenrirUnbound**
## Context

The current proposal for implementing a build timeout requires that tear-down commands always be executed when the build times out. This refactor splits the set of commands up as such.

## Objective

Split the user-defined steps and the tear-down steps. Process the user-defined commands first, then process the tear-down steps.

### CODE REVIEW

**Comment by d2lam** [comment]
Please DNM this yet. There is a bug in the previous launcher PR.

**Comment by FenrirUnbound** [comment]
> Please DNM this yet. There is a bug in the previous launcher PR.

Why does that matter? The previous PR doesn't affect the executor module

**Comment by d2lam** [comment]
@FenrirUnbound Because we wouldn't be able to pull this in. You will not be able to start any builds: https://beta.cd.screwdriver.cd/pipelines/93/builds/8302

**Comment by FenrirUnbound** [comment]
> @FenrirUnbound Because we wouldn't be able to pull this in. You will not be able to start any builds: https://beta.cd.screwdriver.cd/pipelines/93/builds/8302

That's fine, but I fail t see how that affects this particular code change though. 

Whenever there's an infrastructure issue, are we changing _all_ PRs to DNM?

File: executor/executor.go [hunk 1]
```diff
@@ -165,72 +182,67 @@ func Run(path string, env []string, emitter screwdriver.Emitter, build screwdriv
...
-			if err := api.UpdateStepStop(buildID, cmd.Name, code); err != nil {
-				return fmt.Errorf("Updating step stop %q: %v", cmd.Name, err)
-			}
+		if err := api.UpdateStepStop(buildID, cmd.Name, code); err != nil {
+			return fmt.Errorf("Updating step stop %q: %v", cmd.Name, err)
 		}
 
-		// Set first error flag
-		if (firstError == nil) && (cmdErr != nil) {
+		if (firstError == nil) {
```
Comment (tkyi) [inline executor/executor.go]: Missing `&& (cmdErr != nil)` check. Do we no longer need it?

Comment (tkyi) [reply executor/executor.go]: Okay, that makes sense 👍

**Pull request merged**